src/NovelRT/Experimental/Graphics/Vulkan/VulkanGraphicsContext.cpp

```diff
@@ -0,0 +1,519 @@
+
+                    vkUpdateDescriptorSets(GetDevice()->GetVulkanDevice(), 1, &writeDescriptorSet, 0, nullptr);
+                }
+            }
+
+            vkCmdBindDescriptorSets(vulkanCommandBuffer, VK_PIPELINE_BIND_POINT_GRAPHICS,
+                                    pipelineSignature->GetVulkanPipelineLayout(), 0, 1, &vulkanDescriptorSet, 0,
+                                    nullptr);
+        }
+
+        const GraphicsMemoryRegion<GraphicsResource>& indexBufferRegion = primitive->GetIndexBufferRegion();
+        std::shared_ptr<VulkanGraphicsBuffer> indexBuffer =
+            std::dynamic_pointer_cast<VulkanGraphicsBuffer>(indexBufferRegion.GetCollection());
+
+        if (indexBuffer != nullptr)
+        {
+            uint32_t indexBufferStride = primitive->GetIndexBufferStride();
+            VkIndexType indexType = VK_INDEX_TYPE_UINT16;
+
+            if (indexBufferStride != 2)
+            {
+                if (indexBufferStride != 4)
+                {
+                    throw std::out_of_range("An out of range stride value was provided to the index buffer.");
+                }
+
+                indexType = VK_INDEX_TYPE_UINT32;
+            }
+
+            vkCmdBindIndexBuffer(vulkanCommandBuffer, indexBuffer->GetVulkanBuffer(), indexBufferRegion.GetOffset(),
+                                 indexType);
+
+            vkCmdDrawIndexed(vulkanCommandBuffer,
+                             static_cast<uint32_t>(indexBufferRegion.GetSize() / indexBufferStride), 1, 0, 0, 0);
+        }
+        else
+        {
+            vkCmdDraw(vulkanCommandBuffer,
+                      static_cast<uint32_t>(vertexBufferRegion.GetSize() / primitive->GetVertexBufferStride()), 1, 0,
+                      0);
+        }
+    }
+
+    void VulkanGraphicsContext::EndDrawing()
+    {
+        vkCmdEndRenderPass(GetVulkanCommandBuffer());
+    }
+
+    void VulkanGraphicsContext::EndFrame()
+    {
+        VkCommandBuffer commandBuffer = GetVulkanCommandBuffer();
+
+        VkResult endCommandBufferResult = vkEndCommandBuffer(commandBuffer);
+
+        if (endCommandBufferResult != VK_SUCCESS)
+        {
+            throw Exceptions::InitialisationFailureException(
+                "Failed to finalise initialisation for the VkCommandBuffer Instance. This instance cannot be "
+                "submitted.",
+                endCommandBufferResult);
+        }
+
+        VkSubmitInfo submitInfo{};
+        submitInfo.sType = VK_STRUCTURE_TYPE_SUBMIT_INFO;
+        submitInfo.commandBufferCount = 1;
+        submitInfo.pCommandBuffers = &commandBuffer;
+
+        std::shared_ptr<VulkanGraphicsFence> executeGraphicsFence = GetWaitForExecuteCompletionFence();
+
+        // TODO: Might need a fence here? Not sure yet.
+        VkResult queueSubmitResult = vkQueueSubmit(GetDevice()->GetVulkanGraphicsQueue(), 1, &submitInfo,
+                                                   executeGraphicsFence->GetVulkanFence());
```

Comment:
Judging by the function, it looks like we did need a fence here? Or did I misread the TODO?

Reply:
I did indeed end up adding a fence. :)